### PR TITLE
[sparse.csgraph] Fix a bug in dijkstra and johnson algorithm

### LINKED ISSUE #17782: BUG: Segfault in scipy.sparse.csgraph.shortest_path() with v1.10.0
### Describe your issue.

Hello! Since last week’s upgrade from scipy 1.9.3 to 1.10.0, we have encountered a reproducible segmentation fault on running `scipy.sparse.csgraph.shortest_path()` with a particular weighted adjacency matrix and two `method` parameters. The matrix is attached. The problem has been observed with Python 3.9.15 on both macOS 12.6.2/arm64 and Debian 11.6/x86\_64.
[test_case.zip] DELETED

### Reproducing Code Example

```python
import scipy.sparse

mat = scipy.sparse.load_npz("test_case.npz")

# these crash
scipy.sparse.csgraph.shortest_path(mat)
scipy.sparse.csgraph.shortest_path(mat, method="D")
scipy.sparse.csgraph.shortest_path(mat, method="J")

# these don't
scipy.sparse.csgraph.shortest_path(mat, method="FW")
scipy.sparse.csgraph.shortest_path(mat, method="BF")
```

### Error message

```shell
segmentation fault
```

### SciPy/NumPy/Python version information

1.10.0 1.24.1 sys.version_info(major=3, minor=8, micro=15, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by windows-server-2003**
#### Reference issue
Closes #17782

#### What does this implement/fix?
My previous pull request #16696 introduced a bug in FibonacciHeap implementation in sparse/csgraph/_shortest_path.pyx which caused the dijkstra algorithm to crash or run into an infinite loop in some cases.
This pull request provides a fix for the bug as well as brute force tests using random graphs to find and avoid such bugs in the future.

The bug is that I accidentally wrote `node.right_sibling` instead of `node` which broke the linked-list structure.
One of the added tests `test_dijkstra_random` runs `dijkstra` with a lot of random graphs and is expected to find crashes/infinite loops. It indeed does not terminate for the (current) bugged version.
The other one `test_shortest_path_random` runs `shortest_path` with all methods with similar random graphs and checks the distance table/predecessor table are valid and consistent across methods. This does not immediately find any bugs, but I think this helps finding bugs in the future change.
The two tests run together in 3 seconds (when the bug is fixed) on my machine(i5-10600k, Ubuntu 20.04 on VirtualBox), which I hope is acceptable.

I found a small mistake in the specification of `directed` option in another test, so I fixed that too.

I apologize for having contaminated the project with buggy code.
I am still looking for a way to get rid of manual FibonacciHeap implementation in #17019 but still need some time due to the "too many public symbols" error, so I would like this pull request to be reviewed and merged first.

### CODE REVIEW

**Review by Kai-Striega — CHANGES_REQUESTED** [review]
I have verified your fix with both provided matrices in the reference issue. It works, well done!

I'm not a fan of testing random matrices to see if they cause a problem. I don't like randomness in tests in general. What I would recommend instead is including the two matrices under a new data/ directory with a descriptive filename to document what they are there for. Then loading them for the test and testing against them.  You will need to change the `meson.build` file to include this new directory in the build files. Have a look at `scipy/ndimage/tests` for an example of how to do this.

**Comment by perimosocordiae** [comment]
+1, changes look good but we should pin down the matrix that exercised the line with the bug. If you don't want to add extra files, you could also hard-code a test matrix (assuming you find one smaller than roughly 10x10).

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (tylerjereddy) [inline scipy/sparse/csgraph/tests/test_shortest_path.py:210]: I think we can avoid iterating over `t` to call `dijkstra()` over and over here--in fact, I couldn't quite figure out why you were doing that with a fixed seed, but it looks like the seed may not have been passed to `rand`, so you were doing a random search

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (tylerjereddy) [inline scipy/sparse/csgraph/tests/test_shortest_path.py:214]: here's a draft version of the test, with a fixed matrix (no searching), that hangs on `main` and passes on this branch, after doing some brute force searching for a small 10x10 reproducer per CJs suggestion (I also modernized the rng infra while I was there):

```diff
--- a/scipy/sparse/csgraph/tests/test_shortest_path.py
+++ b/scipy/sparse/csgraph/tests/test_shortest_path.py
@@ -200,18 +200,16 @@ def test_dijkstra_min_only_random(n):
 
 
 # ensure dijkstra does not crash or run into an infinite loop
-@pytest.mark.parametrize('n, t',
-                         ((3, 300),
-                          (10, 100),
-                          (100, 10),
-                          (500, 1)))
-def test_dijkstra_random(n, t):
-    np.random.seed(1234)
-    for testcase_index in range(t):
-        data = scipy.sparse.rand(n, n, density=2.0 / n, format='lil',
-                                 dtype=np.float64)
-        data.setdiag(np.zeros(n, dtype=np.bool_))
-        dijkstra(data, directed=True, return_predecessors=True)
+def test_dijkstra_random():
+    n = 10
+    # this random seed was carefully chosen in a search
+    # for reproducers of gh-17782
+    rng = np.random.default_rng(3443)
+    data = scipy.sparse.rand(n, n, density=2.0 / n, format='lil',
+                             dtype=np.float64,
+                             random_state=rng)
+    data.setdiag(np.zeros(n, dtype=np.bool_))
+    dijkstra(data, directed=True, return_predecessors=True)
```

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (windows-server-2003) [reply scipy/sparse/csgraph/tests/test_shortest_path.py:3]: Should I also replace existing `assert_almost_equal` in the other tests in the file too?

Comment (WarrenWeckesser) [reply scipy/sparse/csgraph/tests/test_shortest_path.py:3]: @windows-server-2003, you don't have to replace existing occurrences of those test functions in code that you are not contributing.  Just be sure to use the recommended ones in new code.

This is a frequently occurring issue in pull requests.  I think we (the SciPy devs) should consider doing a bulk replacement of all the discommended (i.e. no longer recommended but not actually deprecated) functions, to avoid the natural tendency of contributors to copy the existing tests.

Comment (windows-server-2003) [reply scipy/sparse/csgraph/tests/test_shortest_path.py:3]: Thanks, understood.

Comment (WarrenWeckesser) [reply scipy/sparse/csgraph/tests/test_shortest_path.py:3]: > should we perhaps prefer `assert_allclose` to `assert_almost_equal` in new test code? 

Yes, we decided that some time ago, and made note of that decision in the "missing bits" developer documentation: https://docs.scipy.org/doc/scipy/dev/missing-bits.html#test-functions-from-numpy-testing

@windows-server-2003, we don't expect all new contributors to be familiar with that document, but now that you have a couple PRs in progress, it would be worthwhile to browse through it.

**Review by Kai-Striega — COMMENTED** [review]
Test looks good now, thanks for taking the time to contribute

**Review by Kai-Striega — APPROVED** [review]

File: scipy/sparse/csgraph/tests/test_shortest_path.py
Comment (perimosocordiae) [inline scipy/sparse/csgraph/tests/test_shortest_path.py:205]: Maybe I'm paranoid, but I don't fully trust this to not change on us a few years down the line. Let's actually hard-code it:

```python
n = 10
indices = [0, 4, 4, 5, 7, 9, 0, 6, 2, 3, 7, 9, 1, 2, 9, 2, 5, 6]
indptr = [0, 0, 2, 5, 6, 7, 8, 12, 15, 18, 18]
data = [0.33629, 0.40458, 0.47493, 0.42757, 0.11497, 0.91653, 0.69084, 0.64979, 0.62555, 0.743  , 0.01724, 0.99945, 0.31095, 0.15557, 0.02439, 0.65814, 0.23478, 0.24072]
graph = scipy.sparse.csr_matrix((data, indices, indptr), shape=(n, n))
dijkstra(graph, directed=True, return_predecessors=True)
```

(some line-wrapping needed, but that's the idea)

**Comment by tylerjereddy** [comment]
@windows-server-2003 let me know if you need a hand addressing the suggestion from CJ--I think this is close now and a hang/segfault is certainly something I'd be keen to backport for a bugfix release.

**Comment by Kai-Striega** [comment]
The only concern that I still have is that there are two cases; one where it hangs and one where it segfaults. This PR is only adding one test matrix but fixes both cases, should we try to find an example of the other case?

**Comment by tylerjereddy** [comment]
> should we try to find an example of the other case?

This was much harder to do than for the hanging case, but I've cut the segfault case down from 270 nodes or so to 22 nodes. The search for segfault reproducers gets increasingly difficult as the number of nodes drops, but maybe we can live with this? Or maybe CJ can see a pattern now that the data is small enough... but not sure it is worth dissecting too much more. I would point out the presence of multiple duplicated floating point "distances."

```python
from io import StringIO
from scipy.io import mmread
from scipy.sparse.csgraph import dijkstra


def test_gh_17782_segfault(tmp_path):
    text = """%%MatrixMarket matrix coordinate real general
                84 84 22
                2 1 4.699999809265137e+00
                6 14 1.199999973177910e-01
                9 6 1.199999973177910e-01
                10 16 2.012000083923340e+01
                11 10 1.422000026702881e+01
                12 1 9.645999908447266e+01
                13 18 2.012000083923340e+01
                14 13 4.679999828338623e+00
                15 11 1.199999973177910e-01
                16 12 1.199999973177910e-01
                18 15 1.199999973177910e-01
                32 2 2.299999952316284e+00
                33 20 6.000000000000000e+00
                33 32 5.000000000000000e+00
                36 9 3.720000028610229e+00
                36 37 3.720000028610229e+00
                36 38 3.720000028610229e+00
                37 44 8.159999847412109e+00
                38 32 7.903999328613281e+01
                43 20 2.400000000000000e+01
                43 33 4.000000000000000e+00
                44 43 6.028000259399414e+01
    """
    data = mmread(StringIO(text))
    dijkstra(data, directed=True, return_predecessors=True)
```

Segfaults near-instantly on `main`, passes in `0.17` seconds on this feature branch.

**Comment by Kai-Striega** [comment]
Thanks, @tylerjereddy - I think that should be enough to get this PR merged

**Comment by Kai-Striega** [comment]
@windows-server-2003 do you need some help working on this? I'd like to get this into the next bugfix release

**Comment by tylerjereddy** [comment]
I think I'll go ahead and push in my testing suggestion, I'd also like to see this in `1.10.1` if the CI clears up.

**Comment by tylerjereddy** [comment]
I pushed in the hardcoded test cases/adjustments for reproducing the hang and segfault, from CJ and I. I also rebased since the diff is pretty small and it may help clear up CI.

**Comment by tylerjereddy** [comment]
CI was perfectly green, the bug fix is a one-linear, and I think we basically have 2 core dev approvals now that the tests have been hardcoded for reproduction of the hang and segfault, respectively.

I squash-merged to facilitate backporting, and tried to do the co-authored commit stuff correctly for folks reviewing/suggesting stuff.

**Pull request merged**
